src/v/raft/replicate_batcher.cc

```diff
@@ -28,29 +28,33 @@ using namespace std::chrono_literals; // NOLINT
 replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
   : _ptr(ptr)
   , _max_batch_size(cache_size) {
-    _flush_timer.set_callback([this] {
-        (void)ss::with_gate(_ptr->_bg, [this] {
-            // background block further caching too
-            return _lock.with([this] { return flush(); });
-        }).handle_exception_type([this](const ss::gate_closed_exception&) {
-            vlog(
-              _ptr->_ctxlog.debug,
-              "Gate closed while flushing replicate requests");
-        });
+    _flush_timer.set_callback([this] { dispatch_background_flush(); });
+}
+
+void replicate_batcher::dispatch_background_flush() {
+    (void)ss::with_gate(_ptr->_bg, [this] {
+        // background block further caching too
+        return _lock.with([this] { return flush(); });
+    }).handle_exception_type([this](const ss::gate_closed_exception&) {
+        vlog(
+          _ptr->_ctxlog.debug, "Gate closed while flushing replicate requests");
     });
 }
 
 ss::future<result<replicate_result>>
 replicate_batcher::replicate(model::record_batch_reader&& r) {
-    return _lock
-      .with(
-        [this, r = std::move(r)]() mutable { return do_cache(std::move(r)); })
-      .then([this](item_ptr i) {
-          if (_pending_bytes < _max_batch_size || !_flush_timer.armed()) {
-              _flush_timer.rearm(clock_type::now() + 4ms);
-          }
-          return i->_promise.get_future();
-      });
+    return do_cache(std::move(r)).then([this](item_ptr i) {
```

Comment:
i think that the commit message should explain why `do_cache` no longer needs to be protected by `_lock`.

Reply:
I didn't want to include removing lock in the commit, good catch.